pyro/contrib/easyguide/easyguide.py

```diff
@@ -0,0 +1,267 @@
+        """
+        raise NotImplementedError
+
+    def init(self, site):
+        """
+        Model initialization method, may be overridden by user.
+
+        This should input a site and output a valid sample from that site.
+        The default behavior is to draw a random sample::
+
+            return site["fn"]()
+
+        For other possible initialization functions see
+        http://docs.pyro.ai/en/stable/contrib.autoguide.html#module-pyro.contrib.autoguide.initialization
+        """
+        return site["fn"]()
+
+    def __call__(self, *args, **kwargs):
+        """
+        Runs the guide. This is typically used by inference algorithms.
+        """
+        if self.prototype_trace is None:
+            self._setup_prototype(*args, **kwargs)
+        result = self.guide(*args, **kwargs)
+        self.plates.clear()
+        return result
+
+    def plate(self, name, size=None, subsample_size=None, subsample=None, *args, **kwargs):
+        """
+        A wrapper around :class:`pyro.plate` to allow `EasyGuide` to
+        automatically construct plates. You should use this rather than
+        :class:`pyro.plate` inside your :meth:`guide` implementation.
+        """
+        if name not in self.plates:
+            self.plates[name] = pyro.plate(name, size, subsample_size, subsample, *args, **kwargs)
+        return self.plates[name]
+
+    def group(self, match=".*"):
+        """
+        Select a :class:`Group` of model sites for joint guidance.
+
+        :param str match: A regex string matching names of model sample sites.
+        :return: A group of model sites.
+        :rtype: Group
+        """
+        if match not in self.groups:
+            sites = [site
+                     for name, site in self.prototype_trace.iter_stochastic_nodes()
+                     if re.match(match, name)]
+            if not sites:
+                raise ValueError("EasyGuide.group() pattern {} matched no model sites"
+                                 .format(repr(match)))
+            self.groups[match] = Group(self, sites)
+        return self.groups[match]
+
+    def map_estimate(self, name):
+        """
+        Construct a maximum a posteriori (MAP) guide using Delta distributions.
+
+        :param str name: The name of a model sample site.
+        :return: A sampled value.
+        :rtype: torch.Tensor
+        """
+        site = self.prototype_trace.nodes[name]
+        fn = site["fn"]
+        param_name = "auto_{}".format(name)
+        init_value = None
+        if param_name not in pyro.get_param_store():
+            init_value = site["value"].detach()
+        with ExitStack() as stack:
+            for frame in site["cond_indep_stack"]:
+                plate = self.plate(frame.name)
+                if plate not in runtime._PYRO_STACK:
+                    stack.enter_context(plate)
+                elif init_value is not None and plate.subsample_size < plate.size:
```

Comment:
`elif init_value is not None and plate.subsample_size < plate.size` when does this happen?

Reply:
- `init_value not None` happens on the first guide call.
- `plate.subsample_size < plate.size` happens whenever the plate is subsampled.
If both happen, then we need to adjust the `init_value` from subsampled shape to full shape; `pyro.plate` will then subsample back down.